include/circular_queue.h

```diff
@@ -0,0 +1,58 @@
+#ifndef QUEUE_GUARD
+#define QUEUE_GUARD
+
+#include <condition_variable>
+#include <mutex>
+#include <utility>
+
+struct queue_elm {
+	bool dirty;    // is this queue element yet to be processed by sketching (if so do not overwrite)
+	uint32_t size; // the size of this data element
+	char *data;    // the data element which goes in the queue
+};
+
+/*
+ * A circular queue of data elements.
+ * Used in the bufferTree to place leaf data which is ready to be processed.
+ * Has a finite size and will block operations which would either empty or
```

Comment:
why do we block ops that would empty?

Reply:
I think this is poorly worded. I'll fix that. The idea is that peek is blocked when empty and push is blocked when full.